### PR TITLE
Openshift template

### PR DESCRIPTION
**Pull request by maneta**
This PR introduces: 

* Openshift files regarding the entire deployments.
* Lua Rover integration
* Prometheus Metrics Endpot

I have avoided using templates where it was possible. Some files like the secrets have been written as templates because it needs parameters.

### Prometheus Metrics

The metrics have been inspired by this [document](https://www.datadoghq.com/blog/how-to-monitor-nginx/)

### CODE REVIEW

File: .s2i/environment [hunk 1]
```diff
@@ -1 +1,2 @@
+LUA_PATH=./lua_modules/share/lua/5.1/?.lua;./lua_modules/share/lua/5.1/?/init.lua;/usr/local/openresty/luajit/share/lua/5.1/?.lua;./lua_modules/share/lua/5.1/nginx/?.lua;;
```
Comment (mikz) [inline .s2i/environment]: This should no longer be needed with `rover2` image.

File: nginx/main.conf.liquid [hunk 2]
```diff
@@ -10,14 +10,21 @@ events {
...
+  lua_shared_dict prometheus_metrics 10M;
 
   lua_code_cache {{ lua_code_cache }};
 
   lua_package_path "{{ lua_path }};;";
   lua_package_cpath "{{ lua_cpath }};;";
 
   init_by_lua_block {
+    local prometheus_metrics = ngx.shared.prometheus_metrics
+    prometheus = require("prometheus").init("prometheus_metrics")
```
Comment (mikz) [inline nginx/main.conf.liquid]: This should be `require('nginx.prometheus')`

File: openshift/05-service.yml [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+    - name: http-cors-proxy
+      protocol: TCP
+      port: 8080
+      targetPort: 8080
+  selector:
+    app: cors-proxy
+  type: ClusterIP
+  sessionAffinity: None
+status:
+  loadBalancer: {}
```
Comment (mikz) [inline openshift/05-service.yml]: Those fields are not needed to be here and can be removed.

File: openshift/06-prometheus-service.yml [hunk 4]
```diff
@@ -0,0 +1,22 @@
+apiVersion: v1
+kind: Service
+metadata:
+  name: cors-proxy-monitoring
+  creationTimestamp: null
```
Comment (mikz) [inline openshift/06-prometheus-service.yml]: We also don't need this here.

File: openshift/07-route.yml [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+  name: "cors-proxy-route"
+objects:
+- apiVersion: v1
+  kind: Route
+  metadata:
+    name: cors-proxy
+    labels:
+      app: cors-proxy
+  spec:
+    host: ${ENVIRONMENT}cors-proxy.${WILDCARD_DOMAIN}
```
Comment (mikz) [inline openshift/07-route.yml]: Wouldn't be better to leave this empty and let openshift to figure it out? OpenShift can know its public domain and create unique host per project/route.

File: src/cors-proxy/whitelist.lua [hunk 6]
```diff
@@ -4,6 +4,7 @@ local resty_url = require('resty.url')
 local resty_resolver = require('resty.resolver')
 local round_robin = require('resty.balancer.round_robin')
 local lrucache = require('resty.lrucache')
+local prometheus = require("prometheus").init("prometheus_metrics")
```
Comment (mikz) [inline src/cors-proxy/whitelist.lua]: `'nginx.prometheus'`

Comment (mikz) [reply src/cors-proxy/whitelist.lua]: And this is pretty useless, the variable is not used.

File: openshift/07-route.yml (hunk 5 above)
Comment (maneta) [reply openshift/07-route.yml]: yes but our clusters right know does not have this issue solved yet. So, right now, openshift would create a local route which you cant edit the hostname and we would need to create a new route by hand.

I would leave this as it is today and once we have the infra issue solved we could rollback and leave the host creation to Openshift.

Another solution would bet the entire host as a parameter...

What do you think?

Comment (mikz) [reply openshift/07-route.yml]: I'd go with the whole host as a param.

File: nginx/main.conf.liquid [hunk 7]
```diff
@@ -10,14 +10,21 @@ events {
...
   lua_shared_dict whitelist 10m;
+  lua_shared_dict prometheus_metrics 10M;
 
   lua_code_cache {{ lua_code_cache }};
 
   lua_package_path "{{ lua_path }};;";
   lua_package_cpath "{{ lua_cpath }};;";
 
   init_by_lua_block {
+    local prometheus_metrics = ngx.shared.prometheus_metrics
```
Comment (mikz) [inline nginx/main.conf.liquid]: This is not used.

**Review by mikz — APPROVED** [review]

File: nginx/main.conf.liquid (hunk 7 above)
Comment (maneta) [reply nginx/main.conf.liquid]: I thought it was more efficient declaring local variables in Lua... Well looks like I was wrong :)

File: openshift/07-route.yml (hunk 5 above)
Comment (maneta) [reply openshift/07-route.yml]: Ok!

**Pull request merged**